src/Corrade/Utility/DebugStl.h

```diff
@@ -103,9 +103,32 @@ template<class ...Args> Debug& operator<<(Debug& debug, const std::tuple<Args...
 
 namespace Implementation {
 
+template<typename T>
+typename std::add_lvalue_reference<T>::type declare_lvalue_reference() noexcept;
+
+template<typename T>
+struct has_ostream_operator {
+    private:
+        template<typename C>
+        static constexpr auto check(C *) ->
+        typename std::is_same<
+            decltype(declare_lvalue_reference<std::ostream>() << std::declval<C>()),
+            std::add_lvalue_reference<std::ostream>::type
+        >::type;
+
+        template<typename>
+        static constexpr std::false_type check(...);
+
+    public:
+        static constexpr bool value = decltype(check<T>(nullptr))::value;
+};
```

Comment:
There's a (admittedly weirdly named) [CORRADE_HAS_TYPE(className, typeExpression)](https://doc.magnum.graphics/corrade/TypeTraits_8h.html#a19ce628bb598766db0fffb24c19fc9de) macro that already contains the above functionality, but your variant seems to be a lot simpler than what I stole from somewhere back in the day :) If I fix the macro to have the second parameter variadic (so the comma doesn't break it), I *think* we could then reduce the above to be just this (containing the potentially hard-to-understand template magic in a single place):

```cpp
CORRADE_HAS_TYPE(HasOstreamOperator, typename std::is_same<
    decltype(declareLvalueReference<std::ostream>() << std::declval<C>()),
    std::add_lvalue_reference<std::ostream>::type
>::type)
```

(Also adapting the naming to what Corrade uses for type traits.) I hope I didn't overlook something that would make this impossible tho.

Reply:
`CORRADE_HAS_TYPE()` is fixed to allow this in 8d4eb5bc333d8b8e58e5bbbc9946999f17f1648d, and I think it would work with a minor change:

```cpp
CORRADE_HAS_TYPE(HasOstreamOperator, typename std::enable_if<std::is_same<
    decltype(declareLvalueReference<std::ostream>() << std::declval<C>()),
    std::add_lvalue_reference<std::ostream>::type
>::value>::type)
```

I assume you have some way to test the infinite recursion already -- can you confirm? :)

Reply:
yes! works perfectly. A very useful macro too